contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1057 @@
+        // lenders with heldToken
+        if (wasForceClosed) {
+            return;
+        }
+
+        uint256 marginPrincipal = getCurrentPrincipalFromMargin();
+
+        accountForClose(principalTotal.sub(marginPrincipal));
+
+        assert(principalTotal == marginPrincipal);
+    }
+
+    // ============ Helper Functions ============
+
+    /**
+     * Updates the state variables at any time. Only does anything after the position has been
+     * closed or partially-closed since the last time this function was called.
+     *
+     * - Increases the available amount in the highest buckets with outstanding principal
+     * - Decreases the principal amount in those buckets
+     *
+     * @param  principalRemoved  Amount of principal closed since the last update
+     */
+    function accountForClose(
+        uint256 principalRemoved
+    )
+        internal
+    {
+        if (principalRemoved == 0) {
+            return;
+        }
+
+        uint256 newRepaidAmount = Margin(DYDX_MARGIN).getTotalOwedTokenRepaidToLender(POSITION_ID);
+        assert(newRepaidAmount.sub(cachedRepaidAmount) >= principalRemoved);
+
+        uint256 principalToSub = principalRemoved;
+        uint256 availableToAdd = newRepaidAmount.sub(cachedRepaidAmount);
+        uint256 criticalBucketTemp = criticalBucket;
+
+        // loop over buckets in reverse order starting with the critical bucket
+        for (
+            uint256 bucket = criticalBucketTemp;
+            principalToSub > 0 && bucket <= criticalBucketTemp; // prevent underflow on bucket
+            bucket--
+        ) {
+            uint256 principalTemp = Math.min256(principalToSub, principalForBucket[bucket]);
+            if (principalTemp == 0) {
+                continue;
+            }
+            uint256 availableTemp = MathHelpers.getPartialAmount(
+                principalTemp,
+                principalToSub,
+                availableToAdd
+            );
+
+            changeAvailable(bucket, availableTemp, true);
+            changePrincipal(bucket, principalTemp, false);
+
+            principalToSub = principalToSub.sub(principalTemp);
+            availableToAdd = availableToAdd.sub(availableTemp);
+
+            criticalBucketTemp = bucket;
+        }
+
+        assert(principalToSub == 0);
+        assert(availableToAdd == 0);
+
+        setCriticalBucket(criticalBucketTemp);
+
+        cachedRepaidAmount = newRepaidAmount;
+    }
+
+    /**
+     * Updates the state variables when a position is increased.
+     *
+     * - Decreases the available amount in the lowest buckets with available token
+     * - Increases the principal amount in those buckets
+     *
+     * @param  principalAdded  Amount of principal added to the position
+     * @param  lentAmount      Amount of owedToken lent
+     */
+    function accountForIncrease(
+        uint256 principalAdded,
+        uint256 lentAmount
+    )
+        internal
```

Comment:
`private` right? also ditto for other functions maybe?

Reply:
done